src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -271,71 +304,194 @@ private async Task WriteDataAsyncAwaited(int streamId, StreamOutputFlowControl f
+            +-+-------------------------------------------------------------+
+        */
         public Task WriteWindowUpdateAsync(int streamId, int sizeIncrement)
         {
             lock (_writeLock)
             {
+                if (_completed)
+                {
+                    return Task.CompletedTask;
+                }
+
                 _outgoingFrame.PrepareWindowUpdate(streamId, sizeIncrement);
-                return WriteFrameUnsynchronizedAsync();
+                WriteHeaderUnsynchronized();
+                var buffer = _outputWriter.GetSpan(4);
+                Bitshifter.WriteUInt31BigEndian(buffer, (uint)sizeIncrement);
+                _outputWriter.Advance(4);
+                return _flusher.FlushAsync();
             }
         }
 
+        /* https://tools.ietf.org/html/rfc7540#section-6.4
+            +---------------------------------------------------------------+
+            |                        Error Code (32)                        |
+            +---------------------------------------------------------------+
+        */
         public Task WriteRstStreamAsync(int streamId, Http2ErrorCode errorCode)
         {
             lock (_writeLock)
             {
+                if (_completed)
+                {
+                    return Task.CompletedTask;
+                }
+
                 _outgoingFrame.PrepareRstStream(streamId, errorCode);
-                return WriteFrameUnsynchronizedAsync();
+                WriteHeaderUnsynchronized();
+                var buffer = _outputWriter.GetSpan(4);
+                BinaryPrimitives.WriteUInt32BigEndian(buffer, (uint)errorCode);
+                _outputWriter.Advance(4);
+
+                return _flusher.FlushAsync();
             }
         }
 
+        /* https://tools.ietf.org/html/rfc7540#section-6.5.1
+            List of:
+            +-------------------------------+
+            |       Identifier (16)         |
+            +-------------------------------+-------------------------------+
+            |                        Value (32)                             |
+            +---------------------------------------------------------------+
+        */
         public Task WriteSettingsAsync(IList<Http2PeerSetting> settings)
         {
             lock (_writeLock)
             {
-                _outgoingFrame.PrepareSettings(Http2SettingsFrameFlags.NONE, settings);
-                return WriteFrameUnsynchronizedAsync();
+                if (_completed)
+                {
+                    return Task.CompletedTask;
+                }
+
+                _outgoingFrame.PrepareSettings(Http2SettingsFrameFlags.NONE);
+                var settingsSize = settings.Count * Http2FrameReader.SettingSize;
+                _outgoingFrame.PayloadLength = settingsSize;
+                WriteHeaderUnsynchronized();
+
+                var buffer = _outputWriter.GetSpan(settingsSize).Slice(0, settingsSize); // GetSpan isn't precise
```

Comment:
How so? I'm wondering if it affects other places where GetSpan is called, especially on irregular numbers such as Http2FrameReader.HeaderLength

Reply:
GetSpan gives you the current buffer if your hint size will fit in it, otherwise it gives you the next buffer. It doesn't slice it down to the size your asked for though. It's also limited to 4kb.